xarray/core/datatree.py

```diff
@@ -93,14 +108,66 @@ def _coerce_to_dataset(data: Dataset | DataArray | None) -> Dataset:
     return ds
 
 
-def _check_for_name_collisions(
-    children: Iterable[str], variables: Iterable[Hashable]
+def _join_path(root: str, name: str) -> str:
+    return root.rstrip("/") + "/" + name
+
+
+def _inherited_dataset(ds: Dataset, parent: Dataset) -> Dataset:
+    return Dataset._construct_direct(
+        variables=parent._variables | ds._variables,
+        coord_names=parent._coord_names | ds._coord_names,
+        dims=parent._dims | ds._dims,
+        attrs=ds._attrs,
+        indexes=parent._indexes | ds._indexes,
+        encoding=ds._encoding,
+        close=ds._close,
+    )
+
+
+def _indented_without_header(text: str) -> str:
+    return textwrap.indent("\n".join(text.split("\n")[1:]), prefix="    ")
+
+
+def _drop_data_vars_and_attrs_sections(text: str) -> str:
+    lines = text.split("\n")
+    outputs = []
+    match = "Data variables:"
+    for line in lines:
+        if line[: len(match)] == match:
+            break
+        outputs.append(line)
+    return "\n".join(outputs)
+
+
+def _check_alignment(
+    path: str,
+    node_ds: Dataset,
+    parent_ds: Dataset | None,
+    children: Mapping[str, DataTree],
 ) -> None:
-    colliding_names = set(children).intersection(set(variables))
-    if colliding_names:
-        raise KeyError(
-            f"Some names would collide between variables and children: {list(colliding_names)}"
-        )
+    if parent_ds is not None:
+        try:
+            align(node_ds, parent_ds, join="exact")
+        except ValueError as e:
+            node_repr = _indented_without_header(repr(node_ds))
+            parent_repr = _indented_without_header(
+                _drop_data_vars_and_attrs_sections(repr(parent_ds))
+            )
+            raise ValueError(
+                f"group {path!r} is not aligned with its parent:\n"
+                f"Group:\n{node_repr}\nParent:\n{parent_repr}"
+            ) from e
```

Comment:
I really like this - any alignment errors are super clear.

It can probably be refactored to integrate a little better with the existing repr-building code, but that can wait.